locator/tablets.cc

```diff
@@ -245,9 +246,26 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
     }
 }
 
-host_id tablet_map::get_primary_replica(tablet_id id) const {
+tablet_replica tablet_map::get_primary_replica(tablet_id id) const {
     const auto info = get_tablet_info(id);
-    return info.replicas.at(size_t(id) % info.replicas.size()).host;
+    auto replicas = boost::copy_range<std::vector<tablet_replica>>(info.replicas);
+    std::sort(replicas.begin(), replicas.end(), [] (const tablet_replica& l, const tablet_replica& r) {
+        return l.host < r.host;
+    });
+    return replicas.at(size_t(id) % replicas.size());
+}
+
+tablet_replica tablet_map::get_primary_replica_within_dc(tablet_id id, const topology& topo, const sstring& dc) const {
+    tablet_replica_set replicas = boost::copy_range<tablet_replica_set>(get_tablet_info(id).replicas | boost::adaptors::filtered([&] (const tablet_replica& r) {
+        return topo.get_datacenter(r.host) == dc;
+    }));
+    if (replicas.empty()) {
+        throw std::runtime_error(format("Cannot get primary replica for tablet_id={} in dc={}: no replicas found in this datacenter", id, dc));
+    }
+    std::sort(replicas.begin(), replicas.end(), [] (const tablet_replica& l, const tablet_replica& r) {
+        return l.host < r.host;
```

Comment:
auto idx = size_t(id) % replicas.size();
std::partial_sort(replicas, replicas.begin() + (idx + 1));

Reply:
interesting idea.
And we can sort in reverse if idx is greater than the size/2

Reply:
Don't we need a stable sort?

The doc for std::partial_sort says:

The order of equal elements is not guaranteed to be preserved. The order of the remaining elements in the range [middle, last) is unspecified.

Reply:
after the partial sort, the elements in [middle, last) are guaranteed to be ordered after [first, middle). and preexisting order of equal elements is not a requirement for us, as replicas will have different hosts. the idea is to reduce complexity from O(n * log n) to O(n * log m), where m = distance from first element to middle one.

Reply:
Yes, I know the idea behind the partial sort.  I have concerns non stable sort would choose the wrong host. 

Assume n1 == n2 in sorting.

before sort: n4 n3 n1 n2

after stable sort:
n1, n2, n3, n4

1) after non stable sort version 1:
n2, n1, n3, n4

2) after non stable sort version 2:
n1, n2, n3, n4

for 1) with idx = 0, we choose n2
for 2) with idx = 0, we choose n1

Reply:
I think you misinterpreted 'The order of the remaining elements in the range [middle, last) is unspecified.'. We're not interested in [middle, last), that's beyond the index to be used (size_t(id) % replicas.size()). The order of elements in the range [first, middle) is not unspecified, they will be sorted.

Reply:
Ah ok it's me who misinterpreted you, it seems. I don't follow this 'n1 == n2' too late for me. will leave the judgment for Benny. If this is not a hot path, optimizating it doesn't matter.

Reply:
Nope, I am not talking about the remaining elements. I am talking about the sorted ones. The sorted ones do not preserve the order for equal elements.

Reply:
However, since we use a unique id for sorting, the equal case will not happen. Not preserving the order is not a problem for this use case.